### PR TITLE
Removing hard flex/bison dependency: only devs need to run the parser

### PR DESCRIPTION
**Pull request by abouteiller**
Removing hard flex/bison dependency: only devs need to run the parser
generators when they change the .y, .l files. Some warnings have been
added to the build system to let developper know that files are newer
than dependents.

- [x] build with bison/flex to builddir by default (if tooling found)
- [x] write back modified files to the version controlled files if .y or .l changed
- [x] update so that dirty checks are done using git status (git doesn't conserve mtime)
- [x] move generated file somewhere they do not conflict with in-source builds
- [x] copy the generated files to the build dir with add_custom_command rather than depending on them
- [x] see if possible to use FLEX_TARGET twice to generate both the build-dir version, and the fallback version (with --noline for the later) (the difficulty is to pass the result of the first COMMAND that runs git status to the second command that does the FLEX_TARGET, as the FLEX_TARGET must be in the caller CMakeLists, yet the execute_process(git) must be in a script)
- [x] do not use compression (git stuff is already compressed, there is no point)
- [x] committing to the .l.c archive (and the .l) requires review from Staff
- [x] a PR that changes the .l without changing the archive will fail CI tests

Signed-off-by: Aurélien Bouteiller <bouteill@icl.utk.edu>

### CODE REVIEW

**Comment by bosilca** [comment]
I can't get this to work, it never print the error message about generated files older than the source. I think this is because the IS_NEWER_THAN is only evaluated at configure time, not during the build process itself.

Also, this is not what we discussed in the past: we said that if flex/bison is detected then nothing changes compare with today, and if not we will use the file provided in the archive. The proposed approach was easier to do, because it could be done during cmake time simply by copying the generated file into the binary file, and keeping everything else as is (such that changing the lex file will indeed trigger the regeneration).

**Comment by abouteiller** [comment]
> I can't get this to work, it never print the error message about generated files older than the source. I think this is because the IS_NEWER_THAN is only evaluated at configure time, not during the build process itself.
> 
Correct, I was looking at a rule to complain at compile time instead, it should be doable but is not super easy.

> Also, this is not what we discussed in the past: we said that if flex/bison is detected then nothing changes compare with today, and if not we will use the file provided in the archive. The proposed approach was easier to do, because it could be done during cmake time simply by copying the generated file into the binary file, and keeping everything else as is (such that changing the lex file will indeed trigger the regeneration).

The advantage is that we do not unintentionally modify the version controlled files with `make clean`, the disadvantage is that we will forget to update the version controlled .y.c files when we do update the .y. An idea of how we can avoid that is welcome, maybe having a pre-commit hook?

File: parsec/utils/keyval_lex.l.c [hunk 1]
```diff
@@ -1,6 +1,6 @@
-#line 1 "/Users/abouteiller/Devel/parsec/dplasma/parsec/parsec/utils/keyval_lex.l.c"
+#line 1 "/Users/abouteiller/Devel/parsec/dplasma/parsec/build/parsec/utils/keyval_lex.l.c"
```
Comment (bosilca) [inline parsec/utils/keyval_lex.l.c:1]: These are absolute path that contain private information. Should not be in the main repo.

**Comment by bosilca** [comment]
At this point it's a complicated piece of code, too complex to my taste and to the extent of the added capabilities.

Moreover, I would not have done it this way. Instead I would exploit the tracking capabilities of cmake, and the flex/bison generators the way they were intended to be used, but copy the developer provided generated files directly into the binary tree during cmake time. I would add a custom command that depends on the generated files, such that every time cmake triggers the regeneration of these files (aka. every time the .[ly] files are altered), we get a chance to react. This custom target will not copy the file back, to avoid automatically polluting the git tree, but instead will enqueue a message to be printed out at the end of the build. My only doubt about this approach is that I do not know how to force a target to be last in the build, without creating unnecessary dependencies to all other targets (which will be messy and will add a lot of changes).

**Comment by abouteiller** [comment]
I can change it to copy the pre-generated to the build tree, that's not very difficult. I can also move the pre-generated somewhere else, like `contrib` to make it clear that they should not be edited. 

Note that the write back script will execute the copy back only when the .y file is modified so there is no 'polluting' of the sources if you don't actually change the .y sources. 

### As for using normal dependency tracking:
Normal tracking of dependencies to produce version controlled files does not work properly because it will add the .y.c files to `make clean` targets , which is another (more systematic and unwanted) way of polluting the git status with random changes (like generating with a different version of Flex after `make clean`). 

Also note that file timestamps are not enforced by git, so the date ordering of .y and .y.c depends in which order git creates the files during checkout, not on commit date or anything like that. Even if we manage to restore `mtimes` for .y.c files (for example we could version control an archive and unpack it, that would restore .y.c file `mtimes`), there would be discrepancies between the build-dir timestamps and the source timestamps (showing the sources as older, while the build products are actually outdated), causing confusion.  Overall it's not clear that normal dependency tracking can be used to trigger 'on-demand' rebuild of version controlled files without odd side effects and randomness, hence using `git status` as a proxy to track changed .y files. 

### As for not automatic write back
I thought about adding a custom target (that is not part of ALL or any other normal byproducts of make) so that we can hand invoke it when we want to update the .y.c version controlled files (eg `ninja update_flex_bison`). However, I do think that if we do not automate the updating of the .y.c files, they will just not get updated (we will forget to do it, even if there is a message) and we will not notice because we develop on systems with good version of `bison` and `flex`.

**Comment by bosilca** [comment]
There will be no issue with the timestamp for 2 reasons:
- git does not care about the timestamp of the file, so you can change it locally during cmake (solution 1)
- the dependency will not use the .c in the source tree, but the one in the build tree. As this file is copied during cmake its timestamp will be higher than the .l file, so the build mechanism will be happy.

The dependencies tracking works great, as proof the 2to3 mechanism we currently have in the tool directory.

The special target to be executed at the end of the build was not for updating files, but for reporting at the end of each build that some files have been updated and the developer might want to copy them back into git. This way the developer is always in charge of when it updates the source tree, and we avoid any unwanted pollution (while still providing the info that something must be done with the generated files).

**Comment by abouteiller** [comment]
I already have modified the thing when we are in the  'no-flex' case and it does indeed work just like 2to3. Push pending. 

It is more difficult when generating the version controlled files (the 'with flex' case), because OUTPUTS of normal rules are added to make clean despite being subject to version control and that's completely unworkable. Maybe that can be avoided if the rule is not part of ALL. Checking on it to see if I can do something that looks more like what you said.

**Comment by abouteiller** [comment]
Ok I changed it, it does mostly what you said:
* we use build-time rules to generate .l.c (and friends) files in build-dir from the .l.c in the version controlled archive. Normal rules will trigger when the .l is modified, and that in turn will double check if the .l is changed from the git index (and abort compilation early rather than use outdated stuff to build). 
* we use normal FLEX_TARGET rules to produce the --noline versions that eventually get version controlled. These get generated into a 'staging area' in the build dir (so that they can get cleaned, they will and is unavoidable)
* the rule that populates the staging area is not implicit, it needs to be called by hand `ninja parsec_pregen_flex_bison`
* the rule that populates the staging also produces an archive (I may remove the tar.gz stage) that is version controlled; the archive essentially considers the staging area, verify that the git status is sane, and copies the staged .l.c files back into the version controlled space (in contrib, so that they don't conflict with in-source builds). This happens only when the explicit call `ninja parsec_pregen_flex_bison` is called, and does nothing if there are stray files and other debris in the build directory.

**Comment by abouteiller** [comment]
Ok I changed it, it does mostly what you said:
* we use build-time rules to generate .l.c (and friends) files in build-dir from the .l.c in the version controlled archive. Normal rules will trigger when the .l is modified, and that in turn will double check if the .l is changed from the git index (and abort compilation early rather than use outdated stuff to build). 
* we use normal FLEX_TARGET rules to produce the --noline versions that eventually get version controlled. These get generated into a 'staging area' in the build dir (so that they can get cleaned, they will and is unavoidable)
* the rule that populates the staging area is not implicit, it needs to be called by hand `ninja parsec_pregen_flex_bison`
* the rule that populates the staging also produces an archive (I may remove the tar.gz stage) that is version controlled; the archive essentially considers the staging area, verify that the git status is sane, and copies the staged .l.c files back into the version controlled space (in contrib, so that they don't conflict with in-source builds). This happens only when the explicit call `ninja parsec_pregen_flex_bison` is called, and does nothing if there are stray files and other debris in the build directory.

**Comment by bosilca** [comment]
Please squash. I will approve and merge after.

File: cmake_modules/pregen_flex_bison.cmake [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+# Missing tool mode: FLEX not found or BISON not found, just check that sources have not been
+# tampered with, if possible (i.e., this is a git worktree)
+if(missing)
+  if(NOT source OR NOT srcdir)
+  # Script Argument errors
+  message(FATAL_ERROR "Script called incorrectly, you need to set -Dsource=xyz.l and -Dsrcdir=\${PROJECT_SOURCE_DIR}")
+  endif()
+
+  execute_process(
```
[the remark this answers is not in the document]
Comment (abouteiller) [reply cmake_modules/pregen_flex_bison.cmake:14]: Someone may still be editing the .y files after expanding a tar.gz, but the most common case is that we will printout some useless 'warning'. I'll comment this out.

**Review by bosilca — APPROVED** [review]

**Pull request merged**